pkg/scheduler/framework/framework.go

```diff
@@ -143,8 +174,738 @@ func (f *framework) EventRecorder() record.EventRecorder {
+	// * obsolete bindings, i.e., bindings that are scheduled in accordance with an out-of-date
+	//   (i.e., no longer active) scheduling policy snapshot; it may or may have been cleared for
+	//   processing by the dispatcher; and
+	// * dangling bindings, i.e., bindings that are associated with a cluster that is no longer
+	//   in a normally operating state (the cluster has left the fleet, or is in the state of leaving),
+	//   yet has not been marked as deleting by the scheduler; and
+	// * deleted bindings, i.e, bindings that are marked for deletion in the API server, but have not
+	//   yet been marked as deleting by the scheduler.
+	//
+	// Note that bindings marked as deleting are ignored by the scheduler (unless they are already marked
+	// for deletion and still has the scheduler cleanup finalizer), as they
+	// are irrelevant to the scheduling cycle.
+	active, creating, obsolete, dangling, deleted := classifyBindings(policy, bindings, clusters)
+
+	// If a binding has been marked for deletion yet still has the scheduler cleanup finalizer,
+	// remove the finalizer from it to clear it for GC.
+	if err := f.removeSchedulerCleanupFinalizerFrom(ctx, deleted); err != nil {
+		klog.ErrorS(err, "failed to remove scheduler cleanup finalizer from deleted bindings", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+		return ctrl.Result{}, err
+	}
+
+	// Mark all dangling bindings as deleting.
+	if err := f.markAsDeletingFor(ctx, dangling); err != nil {
+		klog.ErrorS(err, "failed to mark dangling bindings as deleting", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+		return ctrl.Result{}, err
+	}
+
+	// Check if the scheduler should downscale, i.e., mark some creating/active bindings as deleting and/or
+	// clean up all obsolete bindings right away.
+	//
+	// Normally obsolete bindings are kept for cross-referencing at the end of the scheduling cycle to minimize
+	// interruptions caused by scheduling policy change; however, in the case of downscaling, they can be removed
+	// right away.
+	//
+	// To summarize, the scheduler will only downscale when
+	//
+	// * the scheduling policy is of the PickN type; and
+	// * currently there are too many selected clusters, or more specifically too many creating and active bindings
+	//   in the system; or there are exactly the right number of selected clusters, but some obsolete bindings still linger
+	//   in the system.
+	act, downscaleCount := shouldDownscale(policy, numOfClusters, len(active)+len(creating), len(obsolete))
+
+	// Downscale if needed.
+	//
+	// To minimize interruptions, the scheduler picks creating bindings first, and then
+	// active bindings; when processing active bindings, the logic prioritizes older bindings, i.e., bindings with
+	// smaller CreationTimestamp (assuming monotonic clock in the system).
+	//
+	// This step will also mark all obsolete bindings (if any) as deleting right away.
+	if act {
```

Comment:
good candidate to move out

Reply:
looks like variable act is only used once here, why not just  do

 if act = ; act {
}

Reply:
Fixed. I will refactor this part a little bit more.